src/game-logic/reducers/changeCowBreedingPenResident.js

```diff
@@ -3,26 +3,30 @@ import { COW_GESTATION_PERIOD_DAYS } from '../../constants'
 /**
  * @param {farmhand.state} state
  * @param {farmhand.cow} cow
- * @param {boolean} doAdd If true, cow will be added to the breeding pen. If
+ * @param {boolean} isAdding If true, cow will be added to the breeding pen. If
  * false, they will be removed.
  * @returns {farmhand.state}
  */
-export const changeCowBreedingPenResident = (state, cow, doAdd) => {
-  const { cowBreedingPen } = state
+export const changeCowBreedingPenResident = (state, cow, isAdding) => {
+  const { cowBreedingPen, cowInventory } = state
   const { cowId1, cowId2 } = cowBreedingPen
-  const isPenFull = cowId1 !== null && cowId2 !== null
-  const isCowInPen = cowId1 === cow.id || cowId2 === cow.id
+  const isBreedingPenFull = cowId1 !== null && cowId2 !== null
+  const isCowInBreedingPen = cowId1 === cow.id || cowId2 === cow.id
   let newCowBreedingPen = { ...cowBreedingPen }
 
-  if (doAdd) {
-    if (isPenFull || isCowInPen) {
+  const isCowInInventory = !!cowInventory.find(({ id }) => {
+    return id === cow.id
+  })
+
+  if (isAdding) {
+    if (!isCowInInventory || isBreedingPenFull || isCowInBreedingPen) {
```

Comment:
This change is the crux of this PR. Previously this code assumed that `cow` was in the cow inventory, but there is no guarantee of that being the case in the context of this reducer. Now it will simply bail out if the cow being moved to the breeding pen is not in the inventory.

Reply:
3 conditions is my tipping point for a var or a function, but i also won't hold up merge over this. a `cowCanBeAdded` would sum things up nicely though, and would also allow us to remove a level of nesting. i.e. `if (isAdding && cowCanBeAdded)`

Reply:
This is a great call out! I've simplified this all in https://github.com/jeremyckahn/farmhand/commit/d3d0a0cc66aed8d96248c859ef5381093250d5b2. I also de-nested and cleaned up a bit of other related logic here.

I think your philosophy of using three condition variables as an indicator to refactor into a function is a really good one and I'll keep that in mind going forward!